File: bindings/csharp/AinosSdk/Models/AinosException.cs
namespace AinosSdk.Models;

/// <summary>
/// Base exception for all Ainos SDK errors.
/// </summary>
public class AinosException : Exception
{
    /// <summary>
    /// Optional error code returned by the daemon.
    /// </summary>
    public int? ErrorCode { get; }

    /// <summary>
    /// The raw response type from the daemon, if applicable.
    /// </summary>
    public string? ResponseType { get; }

    public AinosException()
    {
    }

    public AinosException(string message) : base(message)
    {
    }

    public AinosException(string message, Exception innerException) : base(message, innerException)
    {
    }

    public AinosException(string message, int errorCode) : base(message)
    {
        ErrorCode = errorCode;
    }

    public AinosException(string message, int errorCode, string responseType) : base(message)
    {
        ErrorCode = errorCode;
        ResponseType = responseType;
    }
}

File: bindings/java/src/main/java/com/ainos/sdk/models/AinosException.java
package com.ainos.sdk.models;

/**
 * Base exception for all Ainos SDK errors.
 * <p>
 * All SDK exceptions extend this class, allowing callers to catch
 * a single type for all Ainos-related errors.
 */
public class AinosException extends Exception {

    private static final long serialVersionUID = 1L;

    /** Optional error code from the daemon. */
    private final int errorCode;

    /**
     * Constructs a new Ainos exception with the specified detail message.
     *
     * @param message the detail message
     */
    public AinosException(String message) {
        super(message);
        this.errorCode = -1;
    }

    /**
     * Constructs a new Ainos exception with the specified detail message and cause.
     *
     * @param message the detail message
     * @param cause   the root cause
     */
    public AinosException(String message, Throwable cause) {
        super(message, cause);
        this.errorCode = -1;
    }

    /**
     * Constructs a new Ainos exception with a daemon error code and message.
     *
     * @param errorCode the numeric error code from the daemon
     * @param message   the detail message
     */
    public AinosException(int errorCode, String message) {
        super(message);
        this.errorCode = errorCode;
    }

    /**
     * Returns the numeric error code from the daemon, or {@code -1} if not set.
     *
     * @return the error code
     */
    public int getErrorCode() {
        return errorCode;
    }
}